Fixed identifying IF and TF in sys.all_objects and OBJECTPROPERTY

faviansamatha opened the pull request:
### Description

Previously, sys.all_objects mistakenly identified inline table-valued functions (IF), and table-valued functions (TF) as scalar functions (FN). This commit fixes that issue. In addition, this commit fixes the issue for the following properties in OBJECTPROPERTY: IsScalarFunction, IsTableFunction, IsInlineFunction. This commit also fixes misidentifying table functions as scalar functions in SSMS.

Tasks: BABEL-3369, BABEL-3370, BABELFISH-511

Signed-off-by: Favian (Ian) Samatha <ians@bitquilltech.com>


### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

faviansamatha committed:
Fixed identifying IF and TF in sys.all_objects

Previously, sys.all_objects mistakenly identified inline table-valued
functions (IF), and table-valued functions (TF) as scalar functions
(TF). This commit fixes that issue.

Tasks: BABELFISH-483, BABELFISH-508, BABELFISH-511

Signed-off-by: Favian (Ian) Samatha <ians@bitquilltech.com>

faviansamatha committed:
Merge branch 'BABEL_2_X_DEV' into sfavian-dev-sys-all_objects-TF-IF

faviansamatha committed:
Added tests and modified existing tests to use new test framework

Signed-off-by: Favian (Ian) Samatha <ians@bitquilltech.com>

faviansamatha committed:
Modified upgrade test

Signed-off-by: Favian (Ian) Samatha <ians@bitquilltech.com>

faviansamatha committed:
Fixed python validation

Signed-off-by: Favian (Ian) Samatha <ians@bitquilltech.com>

faviansamatha committed:
Merge branch 'BABEL_2_X_DEV' into sfavian-dev-sys-all_objects-TF-IF

faviansamatha committed:
Update expected_create.out

faviansamatha committed:
Merge branch 'BABEL_2_X_DEV' into sfavian-dev-sys-all_objects-TF-IF

faviansamatha committed:
moved upgrade tests

Signed-off-by: Favian (Ian) Samatha <ians@bitquilltech.com>

faviansamatha committed:
reverted some deleted files

Signed-off-by: Favian (Ian) Samatha <ians@bitquilltech.com>

faviansamatha committed:
modified expected values

Signed-off-by: Favian (Ian) Samatha <ians@bitquilltech.com>

faviansamatha committed:
removed extra test from previous version

Signed-off-by: Favian (Ian) Samatha <ians@bitquilltech.com>

faviansamatha committed:
fixed upgrade tests

Signed-off-by: Favian (Ian) Samatha <ians@bitquilltech.com>

faviansamatha committed:
Merge branch 'BABEL_2_X_DEV' into sfavian-dev-sys-all_objects-TF-IF

faviansamatha committed:
Merge branch 'BABEL_2_X_DEV' into sfavian-dev-sys-all_objects-TF-IF

faviansamatha committed:
split up objectproperty tests

Signed-off-by: Favian (Ian) Samatha <ians@bitquilltech.com>

faviansamatha committed:
Added to the upgrade tests

Signed-off-by: Favian (Ian) Samatha <ians@bitquilltech.com>

faviansamatha committed:
fix python expected dependency

Signed-off-by: Favian (Ian) Samatha <ians@bitquilltech.com>

faviansamatha committed:
updated upgrade tests

Signed-off-by: Favian (Ian) Samatha <ians@bitquilltech.com>

faviansamatha committed:
Merge branch 'BABEL_2_X_DEV' into sfavian-dev-sys-all_objects-TF-IF

faviansamatha committed:
added expected files for 13_7 and 13_8

Signed-off-by: Favian (Ian) Samatha <ians@bitquilltech.com>

faviansamatha committed:
Merge branch 'BABEL_2_X_DEV' into sfavian-dev-sys-all_objects-TF-IF

faviansamatha committed:
resetting actions

Signed-off-by: Favian (Ian) Samatha <ians@bitquilltech.com>

faviansamatha committed:
resetting actions

Signed-off-by: Favian (Ian) Samatha <ians@bitquilltech.com>

xhfanhe commented:
Code lgtm, but since this is a quite large change, I would suggest getting another approval just to double check. Thank you!

macwoody commented on contrib/babelfishpg_tsql/sql/sys_views.sql:
```diff
@@ -1292,7 +1306,9 @@ select
 inner join pg_namespace s on s.oid = p.pronamespace
-left join pg_trigger tr on tr.tgfoid = p.oid
+left join pg_trigger tr on tr.tgfoid = p.oid,
+format_type(p.prorettype, null) format_rettype,
```
I feel that we don't need to join these two here since they will be needed only in two cases.

faviansamatha replied:
Done, removed the format_rettype and the func_result

macwoody commented on contrib/babelfishpg_tsql/sql/sys_views.sql:
```diff
@@ -1271,17 +1271,31 @@ select
+        case 
+          when format_rettype = 'trigger'
+            then 'TR'::varchar(2)
+          when func_results LIKE 'TABLE(%' then
```
proretset could replace this string comparison, since that's how "TABLE(" is printed

faviansamatha replied:
Done!

macwoody commented on contrib/babelfishpg_tsql/sql/sys_views.sql:
```diff
@@ -1271,17 +1271,31 @@ select
+            then 'TR'::varchar(2)
+          when func_results LIKE 'TABLE(%' then
+            case 
+              when format_rettype = 'record'
```
I think we could call function here to check if it is 'record' or use 'record''s OID

faviansamatha replied:
Decided to not use the 'record' field due to the indistinguishable case you mentioned.

macwoody commented:
I feel that the implemented approach seems could not distinguish the two cases while. 
See this example. 
```
1> CREATE FUNCTION f2 (@a INT)
2> RETURNS TABLE AS
3> RETURN (SELECT @a + 1 AS col);
4> GO
1> CREATE FUNCTION f3 (@a INT)
2> RETURNS TABLE AS
3> RETURN (SELECT 1 AS COL, @a + 1 as COL2);
4> GO
```
Both of them I thin are SQL inlined table function if I am not wrong, 

```
babelfish_db=# select format_type(prorettype, null), proretset, pg_get_function_result(oid) from pg_proc where proname in ( 'f3' , 'f2');
 format_type | proretset |        pg_get_function_result
-------------+-----------+--------------------------------------
 integer     | t         | TABLE(col integer)
 record      | t         | TABLE("COL" integer, "COL2" integer)

```

faviansamatha committed:
Merge branch 'BABEL_2_X_DEV' into sfavian-dev-sys-all_objects-TF-IF

faviansamatha committed:
Addressed PR comments

Signed-off-by: Favian (Ian) Samatha <ians@bitquilltech.com>

faviansamatha committed:
Modified upgrade scripts

Signed-off-by: Favian (Ian) Samatha <ians@bitquilltech.com>

faviansamatha commented on contrib/babelfishpg_tsql/sql/sys_views.sql:
```diff
@@ -1283,17 +1283,31 @@ select
+            then 'TR'::varchar(2)
+          when p.proretset then
+            case 
+              when t.typtype = 'c'
```
From what I've observed, it seems like Table type function will always return a composite type in pg_types, but inline-table-valued functions will return record or another datatype (like int in the case that it returns a single column)

faviansamatha commented:
> I feel that the implemented approach seems could not distinguish the two cases while. See this example.
> 
> ```
> 1> CREATE FUNCTION f2 (@a INT)
> 2> RETURNS TABLE AS
> 3> RETURN (SELECT @a + 1 AS col);
> 4> GO
> 1> CREATE FUNCTION f3 (@a INT)
> 2> RETURNS TABLE AS
> 3> RETURN (SELECT 1 AS COL, @a + 1 as COL2);
> 4> GO
> ```
> 
> Both of them I thin are SQL inlined table function if I am not wrong,
> 
> ```
> babelfish_db=# select format_type(prorettype, null), proretset, pg_get_function_result(oid) from pg_proc where proname in ( 'f3' , 'f2');
>  format_type | proretset |        pg_get_function_result
> -------------+-----------+--------------------------------------
>  integer     | t         | TABLE(col integer)
>  record      | t         | TABLE("COL" integer, "COL2" integer)
> ```

@macwoody I modified the solution and added these set of tests in sys.all_objects-vu-verify. It should be able to identify these two cases now.

macwoody reviewed (dismissed):
Looks good to me from the view implementation POV

faviansamatha committed:
Merge branch 'BABEL_2_X_DEV' into sfavian-dev-sys-all_objects-TF-IF

faviansamatha committed:
modified objectproperty in case of duplicated object_ids in sys.all_objects

Signed-off-by: Favian (Ian) Samatha <ians@bitquilltech.com>

faviansamatha committed:
Revert "modified objectproperty in case of duplicated object_ids in sys.all_objects"

This reverts commit 64eaa45e8b148e393b59a674da5a249ff793e53c.

faviansamatha committed:
Merge branch 'BABEL_2_X_DEV' into sfavian-dev-sys-all_objects-TF-IF
